Redesign the SIMD macro.

hanliutong opened the pull request:
In this patch, we rename the SIMD macro for more clearer, we also introduce a new macro named `CV_SIMD_OR_VECTOR` instread of `CV_SIMD || CV_SIMD_SCALABLE` to make it more concise

| old name / usage              | new name           |
| ----------------------------- | ------------------ |
| CV_SIMD                       | CV_SIMD            |
| CV_SIMD_SCALABLE              | CV_VECTOR_SIMD     |
| CV_SIMD_SCALABLE_64F          | CV_VECTOR_SIMD_64F |
| CV_SIMD \|\| CV_SIMD_SCALABLE | CV_SIMD_OR_VECTOR  |

And I also think the new macro `CV_SIMD_OR_VECTOR` is also a temporary solution (please correct me if i'm wrong). Since `CV_SIMD` is inherently designed for multiple register widths with size-agnostic code, all the code blocks wrapped in `CV_SIMD` should be compatible with the RVV backend. If we can (and also should) translate/modify all the current CV_SIMD code blocks right now, we do not need the `CV_SIMD_OR_VECTOR` macro anymore. The reason `CV_SIMD_OR_VECTOR` is needed for now is that there are 442 use cases of `CV_SIMD` in 53 files, modifying them would takes a long time.


### Pull Request Readiness Checklist

See details at https://github.com/opencv/opencv/wiki/How_to_contribute#making-a-good-pull-request

- [ ] I agree to contribute to the project under Apache 2 License.
- [ ] To the best of my knowledge, the proposed patch is not based on a code under GPL or another license that is incompatible with OpenCV
- [ ] The PR is proposed to the proper branch
- [ ] There is a reference to the original bug report and related work
- [ ] There is accuracy test, performance test and test data in opencv_extra repository, if applicable
      Patch to opencv_extra has the same branch name.
- [ ] The feature is well documented and sample code can be built with the project CMake

hanliutong commented:
Hello @vpisarev , I have already modified the SIMD macro as we discussed, and I also added the license information in the header of the file. Please review it if you have time.

vpisarev commented:
@hanliutong, do we need CV_VECTOR_SIMD_64F? Is there possibility that 32-bit floating-point arithmetics is available, but 64-bit one is not available?
CV_SIMD_64F has been introduced because in ARMv7 NEON only supported 32-bit floating-point operations.

hanliutong commented:
@vpisarev 
>  Is there possibility that 32-bit floating-point arithmetics is available, but 64-bit one is not available?

Yes, according to [RISC-V V SPEC Chap 18. Standard Vector Extensions](https://github.com/riscv/riscv-v-spec/blob/b6368b3c44d775f8eb01c7ce0ad017db19944aa7/v-spec.adoc#sec-vector-extensions), 64-bit float is not available with Zvl32b, Zve32x and Zve32f. 

However, I think it will  mostly happen on processors for embedded devices, which is not considered in our current implementation, then we always define `CV_VECTOR_SIMD_64F` as 1 (for now). 

Maybe we can keep it for future expansion.

alalek commented:
> CV_VECTOR_SIMD

SIMD is already vectorized data. I see no logic behind adding VECTOR word here.
"SCALABLE" seems misused too. 

May be we want something like VARIABLE_SIMD / SIMD_VARIABLE_VECTOR / SIMD_VVL (variable vector length)?

vpisarev commented:
@alalek, I would insist to use the proposed name. RVV stands for RISC-V Vector extension, ARM SVE stands for Scalable Vector Extension. In general, those instructions should be classified as vector extension. Just using CV_VECTOR would be confusing, because vector has different meanings. CV_VECTOR_SIMD is optimal, because it explicitly mentions vectors and explicitly mentions SIMD, i.e. our SIMD-based HAL.

alalek commented:
> CV_VECTOR_SIMD

It says nothing about variable or fixed size of used SIMD vectors.

hanliutong commented:
Hello @vpisarev, @alalek and @asmorkalov , thank you for the suggestion about the designing of SIMD-related macros.

Since there are a lot of different opinions, I did some work to find the best solution, please see below, and sorry for the long message.

## SIMD and Vector
I would like to give more information firstly about SIMD, Vector, RISC-V Vector, and ARM SVE.

SIMD is the term coming from [Flynn's taxonomy](https://en.wikipedia.org/wiki/Flynn%27s_taxonomy) in 1972 as a set of data-level parallelism computer architectures.

Vector is also a set of data level parallelism (DLP) computer architectures but is not included in Flynn's taxonomy because Vector is released in 1977, a few years later than Flynn's paper. And as Vadim says, vector has different meanings, people usually think of this as a name of data structure, not a name set of ISA.

[Duncan's taxonomy](https://en.wikipedia.org/wiki/Duncan%27s_taxonomy) extends Flynn's taxonomy in order to include vector processes in 1990, it believes that both Vector and SIMD belong to synchronous architectures. However, I think Duncan's taxonomy is not famous as Flynn's taxonomy with rarely used.

John Hennessy and David Patterson give a broader definition of SIMD in their book: Computer Architecture A Quantitative Approach, they define Vector architecture (like RVV),  multimedia SIMD instruction set extensions (Neon, SSE, AVX), and graphics processing units (GPUs) as 3 variations of SIMD (this SIMD is from Flynn's taxonomy, I think).

[ARM SVE](https://developer.arm.com/documentation/102476/0100) is a next-generation SIMD extension to AArch64. Although there is a "vector" in the name, ARM SVE is also a SIMD extension, rather than a vector architecture. More precisely, ARM SVE is a SIMD extension that supports scalable vector length. (I just learned about this part, anyone more familiar with SVE may give more clearer information)

RISC-V Vector is a real vector architecture with scalable vector length.

By the way, the main different between vector architecture and multimedia SIMD instruction set extensions is how the instructions control is implemented (control register like vl and control instruction like setvl is unique in vector) and communication between execution unit and memory unit (elements are processed sequentially in the pipelined unit on the vector architecture).  And whether the register is variable length is not one of the distinguishing features.

## SIMD macros
The above information is given from the architectural perspective, but I think when we discuss SIMD-related macros in OpenCV, we should look at the programming model view.

In my personal opinion, maybe we don't care if the architecture is traditional SIMD or SIMD with predication or pipelined Vector, from the perspective of the Universal Intrinsic designing and programming model, we only care if the length of the vector registers is fixed or variable, in another word, if the instructions are designed for fixed-size vector length only or for variable vector length.

and then let's go back to the starting point and think about the significance of our use of SIMD macros

1. The meaning of the CV_SIMD macro is to indicate that the code in the following code block is data level parallelism by using Universal Intrinsic. This macro does not consider whether the vector is scalable or not, and in practice it refers to fixed length SIMD.
2. We add a RVV back-end to Universal Intrinsic, this is the first scalable back-end, also introduce a new API on Universal Intrinsic. And then we need a new macro for scalable back-end and the new API, which named as "CV_SIMD_SCALABLE"
3. The source code in OpenCV wrapped with CV_SIMD macro need to be modified for scalable back-end by using new API and should be wrapped with CV_SIMD_SCALABLE. And also, since there is a compatibility layer in the original Universal Intrinsic, the modified code can/should also wrapped with CV_SIMD, then we use CV_SIMD || CV_SIMD_SCALABLE.

Accordingly, I find that we need a total of 3 macros. one for scalable vector length, one for fixed vector length, one for both vector length.

## Proposal
Seems the following proposal is the most logical:

| Macro name       | describe                                                     |
| ---------------- | ------------------------------------------------------------ |
| CV_SIMD_FIXED (renamed from CV_SIMD)    | For fixed vector length                                      |
| CV_SIMD_SCALABLE (keeping)| For scalable vector length                                   |
| CV_SIMD         (reuse) | For both vector length, defined as CV_SIMD_SCALABLE \|\| CV_SIMD_FIXED |

However, we need to modify the current CV_SIMD to CV_SIMD_FIXED first, which may cause confusion. And then we need to change it back from CV_SIMD_FIXED to CV_SIMD when some contributor modifies the loop with the new API. But it is reasonable because as I said in the initial comment in this PR, the macro for scalable vector length and for fixed vector length is temporary, when we modify all the SIMD loops with the new API, we only need one macro named CV_SIMD.

Then here is another proposal, which will not require massive modification of the existing CV_SIMD
| Macro name       | describe                                                     |
| ---------------- | ------------------------------------------------------------ |
| CV_SIMD    (keeping)| For fixed vector length                                      |
| CV_SIMD_SCALABLE (keeping)| For scalable vector length                                   |
| CV_SIMD_ANY      (new macro)| For both vector length |

Maybe CV_ANY_SIMD is more fluent.

alalek commented:
**Macro usage**:

We should keep `CV_SIMD` - to keep working state of already existing code (in meaning of the fixed size).
New code for SVE/RVV should use `CV_SIMD_SCALABLE` (or alternative name).
We don't really need `CV_SIMD_ANY` (under any name) as any scalable code should work with "fixed size vectors" by definition (under existed `CV_SIMD` macro). Similar `CV_SIMD || CV_SIMD_SCALABLE` should not be used in optimized code.

"Scalable" code requires more limitations, so optimized "scalable" code is a subset of all optimized SIMD/vector code.

---

**Naming**:

`VECTOR` should be avoided as it confuses developers as "vector" word is already widely used (SIMD register types, functions names through `v_` prefix, etc)

```
grep -Ri vector modules/core/include/opencv2/core/hal/
```

hanliutong commented:
Hello @alalek , I agree with you that we should avoid use "vector" in the name. Not only to avoid confuses, but also because “scalable” is a more appropriate statement.

About the macro usage, I have some different opinions:

> "Scalable" code requires more limitations, so optimized "scalable" code is a subset of all optimized SIMD/vector code.

I think all the code wrapped by `CV_SIMD` macro can be used (by necessary modify) for `CV_SIMD_SCALABLE`. Becasue all the code wrapped in CV_SIMD means it is a wide universal intrinsic code block, which mean it is suitable for 128, 256 and 512 bit. In this sense, it is already variable length and should be compatible with CV_SIMD_SCALABLE. And yes, as you said, there are indeed some codes that require more limitations, but I think all of them are wrapped in CV_SIMD128 and CV_SIMD256, not in CV_SIMD.


> Any scalable code should work with "fixed size vectors" by definition

Yes. But the reason of using CV_SIMD_SCALABLE is we propose the new interface (API) of Universal Intrinsic (and the reason for proposing a new API is that the existing API is not compatible with variable-length architectures). In this case, `CV_SIMD` is not defined in RVV back-end, so we need a macro like `CV_SIMD || CV_SIMD_SCALABLE` for the code written in new API. Please see the sample code below:

|                | Origin Universal Intrinsic | New Universal Intrinsic      |
| -------------- | -------------------------- | ---------------------------- |
| add two vector | va + vb;                   | v_add(va, vb);               |
| get nlanes     | v_float32::nlanes          | Vtraits<v_float32>::vlanes() |

```c++
#if CV_SIMD //origin SIMD loop
	for (int i = 0; i < N; i += v_float32::nlanes )
		vx_load([i]...)
		vc = va + vb; // not work in rvv back-end, need modify
		v_store([i]...)
#endif

#if CV_SIMD || CV_SIMD_SCALABLE //modified SIMD loop
	for (int i = 0; i < N; i += Vtraits<v_float32>::vlanes() )
		vx_load([i]...)
		vc = v_add(va, vb); //Compatible with CV_SIMD macros because of the compatibility layer
		v_store([i]...)
#endif
```
And there is a full example in #22520

For more detiles about rvv backed and the compatibility layer, please refer to [intrin_rvv_scalable.hpp#L18](https://github.com/opencv/opencv/blob/4.x/modules/core/include/opencv2/core/hal/intrin_rvv_scalable.hpp#L18) and [intrin.hpp#L714](https://github.com/opencv/opencv/blob/4.x/modules/core/include/opencv2/core/hal/intrin.hpp#L714)

hanliutong commented:
> If CV_SIMD=1 then CV_SIMD_SCALABLE=1

does it means do some thing like this?
```cpp
#if CV_SIMD
    #define CV_SIMD_SCALABLE 1
#endif
```

This does avoid complex macros, but I propose not to do this becasue of two reasons
1. It may cause confusion. CV_SIMD_SCALABLE will enable in non-variable-length (aka fixed length ) backends. I personally think it doesn't make sense. And code block like following looks like it's specifically design for variable length, but in fact, it is for both length.
```
#if CV_SIMD_SCALABLE 
   for (...)
#endif
```
2. CV_SIMD_SCALABLE is not only used to control the SIMD loop, but also used to control the compatibility layer in Universal Intrinsic. Enable it when CV_SIMD=1 will cause redefinition of some universal intrinsic functions, such as `vx_func` in wide universal intrinsic.

alalek commented:
Logic behind this is simple:
- If you wrote "optimized code" with support of variable vectors/SIMD backend
- then this "optimized code" must work with fixed-size SIMD backend.

(the second case is tested regularly on CI in different ways)

hanliutong commented:
> If you wrote "optimized code" with support of variable vectors/SIMD backend
> then this "optimized code" must work with fixed-size SIMD backend.

Yes, I totally agree with that! In fact, that is why I introduced `CV_SIMD || CV_SIMD_SCALABLE`. 

I would like to clarify the state of these two macros on different backend:
| Backend                             | CV_SIMD | CV_SIMD_SCALABLE |
| ------------------------------------ | ------- | ---------------- |
| RVV (or other scalable target)       | 0       | 1                |
| SSE/AVX/AVX512, ARM Neon, and others | 1       | 0                |
| cpp (without any SIMD extension)     | 0       | 0                |


The optimized code (written by new API) has aready support both variable backend and fixed-size backend, and then with `CV_SIMD || CV_SIMD_SCALABLE` the optimized code is enabled for both backend.

I think we're very close to a consensus, the only difference is that I think a single macro that defined as `CV_SIMD || CV_SIMD_SCALABLE` to make it clearer.

asmorkalov commented:
Hello @alalek and @hanliutong @vpisarev . I want to refresh the discussion and finally drive to some conclusion. I vote for `CV_SIMD_ANY` option proposed by Liutong and discussed on the call:

Macro name | describe
-- | --
CV_SIMD    (keeping) | For fixed vector length
CV_SIMD_SCALABLE (keeping) | For scalable vector length
CV_SIMD_ANY      (new macro) | For both vector length

Or even do not introduce hew shorter macro al all. Pros for the solution:
- No concerns about existing naming and practice in OpenCV. No massive code changes.
- We just add or remove extra condition on place, if the loop is tested in scalable mode or found a bug.

alalek commented:
Just a note, that `#if CV_SIMD_ANY` code guard would require to write VVL/scalable SIMD code inside. This may confuse.

hanliutong commented:
Hello @asmorkalov @alalek @vpisarev , I have updated the code according to the previous discussion. Thanks you all for your suggestions!

I would like to clarify the original idea of introducing a shorter macro: the existing wide universal intrinsic code (guarded with `CV_SIMD`, but not `CV_SIMD_<specific width>`) can be used for the new RVV backend with a few modified(including use `CV_SIMD | CV_SIMD_SCALABLE`). the new SIMD code should also write for all backend, then it will be guarded by `CV_SIMD | CV_SIMD_SCALABLE ` macro. In another word, **both of `CV_SIMD` and `CV_SIMD_SCALABLE` may not be used individually anymore**, so we think introducing a shorter macro is make sense.

asmorkalov commented:
Hello @hanliutong I apologize for large delay. We discussed the naming on core team meeting couple of times and finally have driven to consensus. The team decided not to introduce new macro, because of very low adoption rate for today and no consensus on naming. So,
- please remove new macro from the code. All other changes seams to be useful.
- please update HSV pipeline to use `CV_SIMD || CV_SIMD_SCALABLE` and I'll merge it after Vadim's approve.

hanliutong commented:
Hello @asmorkalov, I have already update this PR and #22520 (HSV) according to your comment.
